src/main/java/com/artipie/rpm/http/RpmUpload.java

```diff
@@ -23,40 +23,125 @@
 
 /**
  * Slice for rpm packages upload.
  *
  * @since 0.8.3
- * @todo #162:30min Finish implementation of RpmUpload
- *  RpmUpload should behave like the defined in
- *  https://github.com/artipie/rpm-adapter/issues/162:
- *  1. Upload HTTP request
- *  Method: PUT
- *  URI: /package.rpm - the name of RPM package
- *  Query params:
- *  override (optional) - if true, override existing package with same name
- *  Body: RPM package data
- *  2. Upload process
- *  User sends RPM package as PUT HTTP request with RPM data in body.
- *  RPM adapter Slice implementation should process this request, store
- *  the package in repository without changing the name.
- *  If package with same name already exist and override query param flag
- *  is not true, then return 409 error.
- *  Artipie Slice returns 202 status on success and trigger metadata update
- *  asynchronously, it should not run multiple metadata updates
- *  simultaneously. Finish the implementation and enable tests in RpmUploadTest.
+ * @todo #253:30min Finish implementation of RpmUpload by implementing parsing of optional
+ *  false by default `override` request parameter: if package with same name already exist and
+ *  `override` query param flag is not true, then return 409 error. Parameter parsing should be
+ *  implemented in `Request` class and tested, in `response` method we need to check parameter
+ *  value and if the file exists in storage. Do not forget about tests.
+ *  For more information please refer to https://github.com/artipie/rpm-adapter/issues/162:
  */
-public class RpmUpload extends Slice.Wrap {
+public final class RpmUpload implements Slice {
 
     /**
-     * New RPM repository HTTP API.
+     * Asto storage.
+     */
+    private final Storage asto;
+
+    /**
+     * Rpm instance.
+     */
+    private final Rpm rpm;
+
+    /**
+     * RPM repository HTTP API.
      *
      * @param storage Storage
      */
     public RpmUpload(final Storage storage) {
-        super(new RpmSlice(storage));
+        this.asto = storage;
+        this.rpm = new Rpm(storage);
+    }
+
+    @Override
+    public Response response(final String line, final Iterable<Map.Entry<String, String>> headers,
+        final Publisher<ByteBuffer> body) {
+        final Request request = new Request(line);
+        return new AsyncResponse(
+            SingleInterop.fromFuture(
+                this.asto.save(
+                    request.file(),
+                    new Content.From(body)
+                ).thenApply(ignored -> true)
+            )
+                .flatMapCompletable(ignored -> this.rpm.batchUpdate(request.repo()))
+                .<Response>andThen(Single.just(new RsWithStatus(RsStatus.ACCEPTED)))
+        );
+    }
+
+    /**
+     * Request line.
+     * @since 0.9
+     */
+    static final class Request {
+
+        /**
+         * RegEx pattern for path.
+         */
+        public static final Pattern PTRN = Pattern.compile("^(?<repo>.*)/(?<rpm>.*\\.rpm)");
```

Comment:
@olenagerasimova you don't need to care about URI path format here, Artipie will automatically remove repo coordinates from it: https://github.com/artipie/artipie/blob/3d2df13cb9c7983cc7b639930e715e3adc5e9511/src/main/java/com/artipie/SliceFromConfig.java#L116

Also, it may be configured differently, and may flat repository structure: `/repo/key` or hierarchical: `/user/repo/key`. So this path in URI is always relative to repository.

Reply:
@g4s8 I do not get it: let's say my path in artipie settings yaml is `/var/artipie/rpm`. I have several rpm repositories inside `/var/artipie/rpm`: `/var/artipie/rpm/one` and `/var/artipie/rpm/two`. How on this adapter side I will know what repo to update?

Reply:
@olenagerasimova you can't update wrong repo in adapter, the `Storage` is separate for different repos. In your example you'll have two **instances** of `RpmSlice` with different `Storage` instances (it's even possible to use file system storage for `/rpm-one` adapter and S3 storage for `/rpm-two`).
Artipie web server find repo name by URI (in `flat` mode it's the first part of path), e.g. for `PUT /rpm-one/some-pacakge.rpm` Artipie instantiates `RpmSlice` and creates `Storage` from `rpm-one.yaml` configuration (it's FS storage at `/var/artipie/rpm/one` in your example), then Artipie removes repo context from URI: `PUT /rpm-one/some-pacakge.rpm` becomes `PUT /some-package.rpm` (for `flat` mode), (full path is available at `X-FullPath` header if needed), so your slice receives only package name request and it has only `rpm-one` storage.